webui/react/src/services/stream/stream.ts

```diff
@@ -0,0 +1,151 @@
+import { forEach, reduce, trimEnd } from 'lodash';
+
+import { decode_keys, KeyCache } from './keyCache';
+import { StreamSpec } from './projects';
+
+import { Streamable, StreamEntityMap } from '.';
+
+// About 60 seconds of auto-retry.
+const backoffs = [0, 1, 2, 4, 8, 10, 10, 10, 15];
+const sleep = (s: number) => new Promise((r) => setTimeout(r, 1000 * s));
+
+type Entity = {
+    keyCache: KeyCache,
+    spec: StreamSpec
+}
+
+export class Stream {
+
+    #ws: WebSocket;
+    #retries: number = 0;
+    #numSyncs: number = 0;
+
+    #entities: Array<Record<Streamable, Entity>> = [];
+    #curEntity: Record<Streamable, Entity> | undefined;
+
+    #syncSent: string | undefined = undefined;
+    #syncComplete: string | undefined = undefined;
+    #syncStarted: string |undefined = undefined;
+
+    constructor(wsUrl: string, onUpsert: (m: Record<string, any>) => void, onDelete: (s: Streamable, a: Array<number>) => void) {
+        this.#ws = new WebSocket(wsUrl);
+        this.#ws.onopen = () => {
+            console.log('Streaming websocket opened!');
+            this.advanceSubscription();
+        };
+        this.#ws.onerror = async (err) => {
+            console.log('Streaming websocket errored: ', err);
+            console.log('# of retries: ', this.#retries);
+            const wait = await this.retry();
+            if (!wait) throw err;
+            this.#ws.dispatchEvent(new Event('open'));
+        };
+
+        this.#ws.onmessage = (event) => {
+
+            const msg = JSON.parse(event.data) as Record<string, any>;
+            if (msg['sync_id']) {
+                if (!msg['complete']) {
+                    this.#syncStarted = msg['sync_id'];
+                } else {
+                    this.#syncComplete = msg['sync_id'];
+                    this.advanceSubscription();
+                }
+            } else {
+                // Ignore all messages between when we send a new subscription and when the
+                // sync-start message for that subscription arrives.  These are the online
+                // updates for a subscription we no longer care about.
+                if (this.#syncSent !== this.#syncStarted) return;
+
+                forEach(msg, (val, k) => {
```

Comment:
This isn't a review per se, but what is this `forEach()` for?  I see it's a function defined in our code, but what's wrong with something more vanilla like:

```ts
Object.entries(msg).forEach(([key, val])) => {})
```

?

Reply:
I don't think js `forEach` would work on `Record` type

Reply:
Sure it will:

```ts
Object.entries(msg).forEach([k, value] => {...})
```

I don't even know what `lodash` is for?  Is it some kind of polyfill?